### PR TITLE
Remove bpftool dependency

### PR DESCRIPTION
**Pull request by tatry**
Also refactor pipeline module and little changes to CLI and API

### CODE REVIEW

File: lib/psabpf_pipeline.c [hunk 1]
```diff
@@ -113,175 +160,179 @@ static int xdp_port_add(__u32 pipeline_id, char *intf)
...
+            fprintf(stderr, "failed to pin map at %s: %s\n", pinned_file, strerror(-ret));
             goto err_close_obj;
         }
     }
 
 err_close_obj:
     bpf_object__close(obj);
 
-    return ret;
+    return -ret;
```
Comment (kmateuszssak) [inline lib/psabpf_pipeline.c:279]: Why minus?

**Review by kmateuszssak — COMMENTED** [review]
I left some comments.

File: CLI/pipeline.h [hunk 2]
```diff
@@ -4,17 +4,15 @@
...
+int do_pipeline_load(int argc, char **argv);
+int do_pipeline_unload(int argc, char **argv);
+int do_pipeline_port_add(int argc, char **argv);
+int do_pipeline_port_del(int argc, char **argv);
 
 static const struct cmd pipeline_cmds[] = {
         {"help",     do_pipeline_help },
-        {"load",     do_load },
-        {"unload",   do_unload },
-        {"add-port", do_port_add },
```
[the remark this answers is not in the document]
Comment (kmateuszssak) [reply CLI/pipeline.h:16]: It is moved

File: CLI/pipeline.c [hunk 3]
```diff
@@ -28,148 +27,149 @@ int do_load(int argc, char **argv)
...
+        fprintf(stderr, "too many arguments\n");
+        return EINVAL;
     }
 
     char *file = *argv;
 
-    psabpf_pipeline_t pipeline;
-    psabpf_pipeline_init(&pipeline);
-    psabpf_pipeline_setid(&pipeline, id);
+    psabpf_context_t pipeline;
```
[the remark this answers is not in the document]
Comment (tatry) [reply CLI/pipeline.c:37]: Maybe it should be, but I decided to preserve original variable name. However, this variable name is not meaningless in this context in my opinion.

File: lib/psabpf_pipeline.c (hunk 1 above)
Comment (tatry) [reply lib/psabpf_pipeline.c:279]: Generally speaking:
- sometimes `libbpf` return error codes in `errno`, they are positive values (especially wrappers for system call);
- sometimes `libbpf` return error codes as a result of function, they are negative values (this case) - see [libbpf source code](https://github.com/libbpf/libbpf/blob/v0.3/src/libbpf.c#L7758-L7766) for reference.

If we assume that our library returns positive error codes, we must make error codes positive from negative ones.

File: lib/psabpf_pipeline.c [hunk 4]
```diff
@@ -113,175 +160,179 @@ static int xdp_port_add(__u32 pipeline_id, char *intf)
...
     char cmd[256];
-    sprintf(cmd, "bpftool net attach xdp pinned %s/%s%d/%s dev %s overwrite",
-            BPF_FS, PIPELINE_PREFIX, pipeline_id, XDP_HELPER_PROG, intf);
-    system(cmd);
     sprintf(cmd, "tc qdisc add dev %s clsact", intf);
     system(cmd);
     memset(cmd, 0, sizeof(cmd));
-    sprintf(cmd, "tc filter add dev %s ingress bpf da fd %s/%s%d/%s",
-            intf, BPF_FS, PIPELINE_PREFIX, pipeline_id, TC_INGRESS_PROG);
+    sprintf(cmd, "tc filter add dev %s ingress bpf da fd %s/%s%u/%s",
```
Comment (osinstom) [inline lib/psabpf_pipeline.c:188]: We can add `TODO` here to use libbpf for installing TC programs, instead of `tc filter`

**Review by osinstom — APPROVED** [review]
It's a big PR, hard to review, but in general LGTM. I'm ok to merge considering that PTF tests are passing with new psabpf. Please merge once you address comments from Mateusz.

**Comment by tatry** [comment]
@kmateuszssak @osinstom I think that I addressed all the issues, but please wait for PTF tests.

**Pull request merged**